contracts/core/EntryPoint.sol

```diff
@@ -113,47 +116,68 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
-                if eq(32,len) {
-                    returndatacopy(0, 0, 32)
-                    innerRevertCode := mload(0)
-                }
-            }
+        bool callSuccess;
+        if (success) {
+            callSuccess = returnData != 0;
+        } else {
+            bytes32 innerRevertCode = returnData;
             if (innerRevertCode == INNER_OUT_OF_GAS) {
                 // handleOps was called with gas limit too low. abort entire bundle.
                 // can only be caused by bundler (leaving not enough gas for inner call)
                 revert FailedOp(opIndex, "AA95 out of gas");
             } else if (innerRevertCode == INNER_REVERT_LOW_PREFUND) {
-                // innerCall reverted on prefund too low. treat entire prefund as "gas cost"
-                uint256 actualGas = preGas - gasleft() + opInfo.preOpGas;
-                uint256 actualGasCost = opInfo.prefund;
-                emitPrefundTooLow(opInfo);
-                emitUserOperationEvent(opInfo, false, actualGasCost, actualGas);
-                collected = actualGasCost;
+                // innerCall reverted on prefund too low.
+                // below we re-calculate the cast cost and refund.
             } else {
+                actualGas = preGas - gasleft();
+                actualGas += _getUnusedGasPenalty(actualGas, opInfo.mUserOp.callGasLimit + opInfo.mUserOp.paymasterPostOpGasLimit);
                 emit PostOpRevertReason(
                     opInfo.userOpHash,
                     opInfo.mUserOp.sender,
                     opInfo.mUserOp.nonce,
                     Exec.getReturnData(REVERT_REASON_MAX_LEN)
                 );
+            }
+        }
 
-                uint256 actualGas = preGas - gasleft() + opInfo.preOpGas;
-                collected = _postExecution(
-                    IPaymaster.PostOpMode.postOpReverted,
-                    opInfo,
-                    context,
-                    actualGas
-                );
+        return _postInnerCall(
+            opInfo,
+            callSuccess,
+            actualGas
+        );
+    }
+
+    /**
+     * call after innerCall executed to call the account's callData and paymaster's postOp.
+     * - calculate the overall gas used and unuased penalty.
+     * - refund payer if needed
+     * - emit UserOperationEvent
+     */
+    function _postInnerCall(
+        UserOpInfo memory opInfo,
+        bool callSuccess,
+        uint256 actualGas)
+    internal virtual returns (uint256 collected) {
+        unchecked {
+            uint256 prefund = opInfo.prefund;
+            uint256 actualGasCost = actualGas * getUserOpGasPrice(opInfo.mUserOp);
+            uint256 refund;
+            if (prefund >= actualGasCost) {
+                refund = prefund - actualGasCost;
+            } else {
+                actualGasCost = prefund;
+                //depending where the over-gas-used was found, we either reverted innerCall or not.
```

Comment:
Instead of writing this comment add a boolean parameter to the `emitPrefundTooLow` - we don't always need it but it will help make the code flow more clear.